Add array support to validator

pjlee11 opened the pull request:
Resolves #404

_Changes the schema.yaml to reflect what parts of the payloads are arrays, now that arrays are supported._

- ~[ ] Tests added for new features~
- [ ] Test engineer approval

Commit:
Enable validation of arrays from schema.yaml

Commit:
split out isArray and typeOf validation

Commit:
WIP schema updates

pjlee11 commented:
Blocked on https://github.com/BBC-News/simorgh/pull/1007

pjlee11 commented:
While testing this we will also cover https://github.com/BBC-News/simorgh/pull/1007

Commit:
appease codeclimate DRY

bcmn reviewed (dismissed):
Looks good. Nice naming, very clear to follow.

ChrisBAshton commented on src/app/dataValidator/helpers/validators/validateType.js:
```diff
@@ -1,13 +1,31 @@
+    log('- Valid type of array');
+  } else {
+    throwError(
+      `Error: Type does not match for '${parentSchemaName}' - expected 'array'`,
```
Worth making this match line 8?

```suggestion
      `Error: Type does not match for '${parentSchemaName}' - expected 'array' got '${typeof dataNode}'`,
```

pjlee11 replied:
This could be misleading because `typeof` on an Array returns `object`. So if there was a regression where this an `array` slipped through the if statement some how it would log `... - expected 'array' got 'object'` even though it is an array.

ChrisBAshton replied:
Oh ok - so for example if the if statement got accidentally changed to `if (!Array.isArray(dataNode)) {` (note the `!`), then if we passed a legitimate array we'd fail the `if` and get the `else`, and it would say `... - expected 'array' got 'object'`

Seems fair. Maybe change to simply `got ${dataNode}`?

pjlee11 replied:
`dataNode` in this case will pretty much always log `[object Object]` but quickly google'd and I think `console.log(JSON.stringify(dataNode))` would give us what we want

ChrisBAshton commented on src/app/dataValidator/helpers/validators/validateType.js:
```diff
@@ -1,13 +1,31 @@
-    if (schemaType === `${typeof dataNode}`) {
-      log(`- Valid type of ${typeof dataNode}`);
+    if (schemaType === 'array') {
+      validateArray(dataNode, parentSchemaName);
```
Actually, not sure this is worth splitting out like this - we're basically repeating the `if (CHECK) log(success) else error(message)` logic in the two subfunctions, when the only thing that changes is the `CHECK`.

I'd propose:

```js
const valid = isValidTypeOf(schemaType, dataNode, parentSchemaName); // THIS function handles the array vs non-array logic
if (valid) {
    log
}
else {
    throwError
}
```

Should make the unit tests nicer too, as you can assert a boolean response from `isValidTypeOf` rather than try to catch errors or test the absence of them.

pjlee11 replied:
This is a nice idea, however I think I'd go for `isValidType`. Part of the reason I split the functions was to be explicit that `typeof` in JS does not match up with `type` in OpenAPI 3.0, due to the fact `typeof [] === 'object'`

ChrisBAshton replied:
So just rename `isValidTypeOf` to `isValidType`? Yeah that seems fine to me 👍

ChrisBAshton reviewed (commented):
Tests and schema look good, implementation could do with a small refactor IMO 👍

Commit:
Refactor validateType and improve logging

- Refactors validateTypeOf and validateArray into a single `isValidType` function which returns a boolean
- Makes the logging and error throwing consistent across the 2 variations
- Improves the logging by dumping the object as a json blob

Commit:
Merge branch 'latest' of github.com:BBC-News/simorgh into add-array-support-to-validator

Commit:
add a test for Array as a property

ChrisBAshton reviewed (approved):
Much nicer 👍 Thanks!

jamesbrumpton commented:
LGTM
